geofence/geofence.go

```diff
@@ -0,0 +1,66 @@
+package echogeofence
+
+import (
+	"net"
+	"net/http"
+	"strings"
+
+	"github.com/circa10a/go-geofence"
+	"github.com/labstack/echo/v4"
+)
+
+// EchoGeofence holds a geofence config
+type EchoGeofence geofence.Config
+
+// config returns a geofence config from go-geofence
+func (e *EchoGeofence) config() *geofence.Config {
+	return &geofence.Config{
+		IPAddress: e.IPAddress,
+		Token:     e.Token,
+		Radius:    e.Radius,
+		CacheTTL:  e.CacheTTL,
+	}
+}
+
+// Middleware looks up the coordinates of a client to see if it's nearby
+func Middleware(echoGeofence *EchoGeofence) echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(c echo.Context) error {
+			// Get client ip, strip port
+			remoteAddr := c.Request().RemoteAddr
+			ipAddress, _, err := net.SplitHostPort(remoteAddr)
+			if err != nil {
+				return err
+			}
+
+			// Ensure not a private ip
+			ip := net.ParseIP(ipAddress)
+			if ip.IsPrivate() || strings.HasPrefix(ipAddress, "::1") {
+				return next(c)
+			}
+
+			// Create new geofence
+			geofence, err := geofence.New(echoGeofence.config())
+			if err != nil {
+				return err
+			}
+
+			// Check if IP address is within defined radius
+			isAllowed, err := geofence.IsIPAddressNear(ipAddress)
+			if err != nil {
+				return err
+			}
+
+			// Return forbidden if not within defined radius
+			if !isAllowed {
+				return c.JSON(http.StatusForbidden, map[string]interface{}{
```

Comment:
You should not use `c.JSON` as it commits the response and does not allow middlewares to handle this case in their own way. c.JSON does not return error so upstream middlewares do not know that we actually ended with error.
Return `echo.ErrForbidden` or `echo.NewHTTPError(http.StatusForbidden, "Forbidden")` instead

Reply:
sure, makes sense. I've made the changes